clippy_lints/src/methods/map_flatten.rs

```diff
@@ -68,16 +71,44 @@ pub(super) fn check<'tcx>(
         },
     };
 
-    let func_snippet = snippet(cx, map_arg.span, "..");
-    let hint = format!(".and_then({})", func_snippet);
+    let mut applicability = Applicability::MachineApplicable;
+    let sugg_span = expr.span.with_lo(map_span.lo());
     let lint_info = format!("called `map(..).flatten()` on an `{}`", caller_type);
-    span_lint_and_sugg(
-        cx,
-        MAP_FLATTEN,
-        expr.span.with_lo(recv.span.hi()),
-        &lint_info,
-        "try using `and_then` instead",
-        hint,
-        Applicability::MachineApplicable,
-    );
+    let help_msgs = [
+        "try replacing `map` with `and_then` instead",
+        "and remove the final `.flatten()`",
+    ];
+    let suggestion = snippet_with_applicability(cx, map_arg.span, "..", &mut applicability);
+    let hint = format!("and_then({})", suggestion);
+    let lines_count = suggestion.lines().count();
+    if lines_count <= MAX_SUGGESTION_HIGHLIGHT_LINES {
+        span_lint_and_sugg(
+            cx,
+            MAP_FLATTEN,
+            sugg_span,
+            &lint_info,
+            help_msgs[0],
+            hint,
+            applicability,
+        );
+    } else {
+        // modify hint string
+        let mid_idx = MAX_SUGGESTION_HIGHLIGHT_LINES / 2;
+        let lines = hint.split('\n').collect::<Vec<_>>();
+        let top_part = lines[..mid_idx].join("\n");
+        let bottom_part = lines[lines_count - mid_idx..].join("\n");
+        // modify span
+        let sm = cx.sess().source_map();
+        let top_line = sm.lookup_line(map_span.lo()).unwrap();
```

Comment:
This `unwrap()` should be avoided in the final version, as it can cause ICEs (Internal Compiler Errors). It's good for prototyping :+1: You can either ignore cases where these errors occur or, maybe better, fall back on the normal suggestion :)